"PSBT Operations" dialog

gwillen opened the pull request:
Add a "PSBT Operations" dialog, reached from the "Load PSBT..." menu item, giving options to sign or broadcast the loaded PSBT as appropriate, as well as copying the result to the clipboard or saving it to a file.

This is based on Sjors' #17509, and depends on that PR going in first. (It effectively replaces the small "load PSBT" dialog from that PR with a more feature-rich one.)

Some notes:
* The way I display status information is maybe unusual (a status bar, rather than messageboxes.) I think it's helpful to have the information in it be persistent rather than transitory. But if people dislike it, I would probably move the "current state of the transaction" info to the top line of the main label, and the "what action just happened, and did it succeed" info into a messagebox.
* I don't really know much about the translation/localization stuff. I put tr() in all the places it seemed like it ought to go. I did not attempt to translate the result of TransactionErrorString (which is shared by GUI and non-GUI code); I don't know if that's correct, but it matches the "error messages in logs should be googleable in English" heuristic. I don't know whether there are things I should be doing to reduce translator effort (like minimizing the total number of distinct message strings I use, or something.)
* I don't really know how (if?) automated testing is applied to GUI code. I can make a list of PSBTs exercising all the codepaths for manual testing, if that's the right approach. Input appreciated.

Sjors commented on src/qt/bitcoingui.cpp:
```diff
@@ -317,6 +317,8 @@ void BitcoinGUI::createActions()
     signMessageAction->setStatusTip(tr("Sign messages with your Bitcoin addresses to prove you own them"));
     verifyMessageAction = new QAction(tr("&Verify message..."), this);
     verifyMessageAction->setStatusTip(tr("Verify messages to ensure they were signed with specified Bitcoin addresses"));
+    m_load_psbt_action = new QAction(tr("Load PSBT..."), this);
```
You should pick a letter to turn into the keyboard shortcut (works on some operating systems, not macOS). E.g. `"&Load PSBT..." if `L` isn't taken.

gwillen replied:
Just went with L -- it's hard for me to check what's going on here since I'm on a Mac, but I don't see it as being taken. (I didn't add a shortcut for loading from clipboard, since L is taken, C is taken, I could use P for PSBT but it seems kind of unnatural.

Sjors commented on src/qt/psbtoperationsdialog.h:
```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2011-2014 The Bitcoin Core developers
```
Nit: fix year

gwillen replied:
Done.

Sjors commented on src/util/error.cpp:
```diff
@@ -14,7 +14,7 @@ std::string TransactionErrorString(const TransactionError err)
             return "No error";
         case TransactionError::MISSING_INPUTS:
-            return "Missing inputs";
+            return "Inputs missing or spent";
```
Cleaning these error messages could probably be its own commit.

gwillen replied:
Yeah, agreed, will fix.

Sjors commented on src/qt/psbtoperationsdialog.cpp:
```diff
@@ -0,0 +1,255 @@
+
+    switch (analysis.next) {
+        case PSBTRole::UPDATER: {
+            showStatus(tr("Transaction is missing some information about inputs."), StatusLevel::WARNING);
```
I wonder if these strings can be moved to a place where the RPC can also use them.

gwillen replied:
That would make sense actually, yeah. I could move them to wherever AnalyzePSBT is. What would the impact of this be on translation issues? I know that tr() is only used for GUI strings, and _() is used for non-GUI strings -- is there precedent for shared strings?

Sjors replied:
I'm not sure actually. @laanwj, @hebasto?

hebasto replied:
@gwillen 
> I know that tr() is only used for GUI strings, and _() is used for non-GUI strings -- is there precedent for shared strings?

Actually, `bilingual_str` type from `util/translation.h` supports both original and translated strings. More details are available in #16362 and #16224.

You could use something like this:

```
#include <util/translation.h>

bilingual_str bs = _("Transaction is missing some information about inputs.");
showStatus(bs.translated, StatusLevel::WARNING);
LogPrintf(bs.original);
```

Sjors commented on src/qt/psbtoperationsdialog.cpp:
```diff
@@ -0,0 +1,255 @@
+    return tx_description.toStdString();
+}
+
+void PSBTOperationsDialog::showStatus(const QString &msg, StatusLevel level) {
```
Should we create a `OperationsDialog` class to make this stuff reusable? (can wait of course)

gwillen replied:
I'm happy to go either way, are there other existing dialogs we'd immediately want to adopt it for? I do like it better than messageboxes, I tried to make it as drop-in as possible a replacement for them.

Sjors commented on src/qt/psbtoperationsdialog.cpp:
```diff
@@ -0,0 +1,255 @@
+
+#include <iostream>
+
+size_t CountPSBTUnsignedInputs(const PartiallySignedTransaction &psbt) {
```
Maybe move this to `psbt.h`?

gwillen replied:
Can do, was dithering about what to do with it.

Sjors reviewed (commented):
Concept ACK

Code at b58e6f70f99d650a8e6ef4a081641a332cf8748e looks mostly alright, but Travis is very sad.

The first time I compiled this on macOS I get a cryptic error when I select the menu item:
```
2020-01-30 11:09:43.908 bitcoin-qt[619:453974] +[NSXPCSharedListener endpointForReply:withListenerName:]: an error occurred while attempting to obtain endpoint for listener 'com.apple.view-bridge': Connection interrupted
```

It worked fine the second time, so perhaps just some unrelated issue due to earlier code I compiled.

<img width="586" alt="Schermafbeelding 2020-01-30 om 11 18 56" src="https://user-images.githubusercontent.com/10217/73440841-52938400-4352-11ea-8a69-17cbe516192d.png">

Ideally disable the sign button if there's nothing we can sign.

> display status information is maybe unusual (a status bar, rather than messageboxes.) 

I like it.

Currently this only changes the Load PSBT flow. Do you also want to apply this dialog to the send / bump fee flow? I'm fine with doing that later too. 

You could also add a menu option to open a PSBT that's currently on the clipboard (e.g. handy when you need to save it as binary).

gwillen commented:
> Ideally disable the sign button if there's nothing we can sign.

Can I determine that without actually trying to sign? I was assuming that (1) I cannot, and (2) I shouldn't try to actually sign for real just to see if it works.  (Or do you mean just disable the button at the point where we are asked to sign, try, and fail?)

> Currently this only changes the Load PSBT flow. Do you also want to apply this dialog to the send / bump fee flow? I'm fine with doing that later too.

I'd rather do that separately, I am trying _really_ hard to cure myself of the disease of biting off way more than I can reasonably do in one go.

> You could also add a menu option to open a PSBT that's currently on the clipboard (e.g. handy when you need to save it as binary).

This I'd be happy to add, since it's a very small change. Just another item below "Load PSBT...", perhaps "Load PSBT from clipboard..."?

Sjors commented:
I indeed wouldn't try actually signing. But you e.g. tell if a wallet is watch-only, and the new ScriptPubKeyManager's can tell if they can sign for things.

gwillen commented:
Aha, and the Travis failure is the same issue we were talking about yesterday in #bitcoin-core-dev, I got scooped by #17261 and need to rebase again and fix errors, will do.

gwillen committed:
[gui] Load PSBT from clipboard

gwillen commented:
Ok, I'm adding a check so that we do not enable the sign button (and we show a helpful note) if the current wallet is privateKeysDisabled().

I would like to use ScriptPubKeyMan to get a better verdict, but it doesn't seem like the functionality I need is quite there. It looks like what I want is something like CWallet::GetSigningProvider, which calls CanProvide on each ScriptPubKeyMan in the wallet; except that the implementation of LegacyScriptPubKeyMan::CanProvide looks like it can return true in cases where we can't actually sign, but "We can still provide some stuff if we have the script". So I can't tell if there's a call I can make that just answers the question "do we believe that this wallet can sign", without myself looping over all our ScriptPubKeyMans, and trying ProduceSignature with each with a DUMMY_SIGNATURE_CREATOR. Obviously if this is true, a better approach would be to enhance the CWallet/ScriptPubKeyMan interface to provide this, but I think that's best left for another PR. (Or if this is already provided and I'm not seeing how, please fill me in.)

gwillen commented:
Actually, here's an approach that seems like it should work -- FillPSBT already does basically all this work for us, since it does all the signing using a HidingSigningProvider, and allows us to tell it not to sign. So _if_ I correctly understand the behavior of HidingSigningProvider, I think this should already tell us how many inputs we could sign, it just doesn't expose the result. If that seems correct to you, I can expose that from FillPSBT, and determine that way where there's anything for us to do here. @Sjors do you know whether this sounds correct?

gwillen committed:
Improve TransactionErrorString messages.

gwillen committed:
FillPSBT: report number of inputs signed (or would sign)

In FillPSBT, optionally report the number of inputs we successfully
signed, as an out parameter. If "sign" is false, instead report the
number of inputs for which GetSigningProvider does not return nullptr.
(This is a potentially overbroad estimate of inputs we could sign.)

gwillen commented:
Updated:
* Disable sign button unless !privateKeysDisabled() and GetSigningProvider() != nullptr (I'm not sure how perfect a proxy that is for "we can sign" but it worked on my positive and negative test cases.)
* Added "load from clipboard" option.
* Added shortcut key in menu.
* Separated error message fixes into their own commit.
* Moved CountPSBTUnsignedInputs into psbt.h.

Did not change:
* Send/bumpfee flow (would rather save for another PR)
* Factoring out PSBT completeness description strings (not a bad idea, but could someone advise me on how that interacts with translation?)
* Factoring out "OperationsDialog" (would probably rather save for another PR, and would like to understand where else we might use it first.)

Other notes:
* Regarding the transient error Sjors got when clicking the menu item, all I can find when googling is that it's some transient issue that shows up for some people on OS X Catalina in multiple apps, with no obvious cause. I have no real insight there.

Please take another look?

promag commented:
Concept ACK.

gwillen commented:
Speculatively attempting to fix the Windows-only fails without a Windows machine to test on. It looks like MinGW defines "ERROR" as a macro, so I can't give an enum constant that name. And it looks like the Visual Studio project file for building Bitcoin-QT is not autogenerated from the Makefiles, but has to be manually updated, so I've attempted to update it.

Sjors commented on src/qt/psbtoperationsdialog.cpp:
```diff
@@ -0,0 +1,265 @@
+            showStatus(tr("Transaction is missing some information about inputs."), StatusLevel::WARN);
+            break;
+        }
+        case PSBTRole::SIGNER: {
```
When all inputs have `"next": "finalizer"` the transaction itself may still be `"next": "signer"`. This seems to be a bug (@achow101 or feature?) in `analyzepsbt`. If it's a bug, we can ignore it for this PR, if it's a feature, we should handle it, though imo it can wait for a followup.

gwillen replied:
I am going to assume this is a bug, so I'm happy to push to a followup. I have been tempted to try to rewrite the flow of AnalyzePSBT, with an eye to simplifying it and making its correctness easier to check.

achow101 replied:
Sounds like a bug

Sjors replied:
https://github.com/bitcoin/bitcoin/pull/18220

Sjors reviewed (approved):
tACK 3e7ca1e (conditional on original PR being merged)

Would be nice to add change detection in the PSBT summary, but it can wait for a followup.

laanwj commented:
Concept ACK!

Sjors commented:
Mmm, it's misinterpreting a 2-of-3 multisig with two signed outputs. Probably related to my [earlier comment](https://github.com/bitcoin/bitcoin/pull/18027#discussion_r377127607). This is more annoying though, because it means I can't broadcast without the third signature.
<img width="586" alt="Schermafbeelding 2020-02-12 om 19 30 41" src="https://user-images.githubusercontent.com/10217/74365443-69b08800-4dce-11ea-9598-6c92961d276e.png">

gwillen commented:
@Sjors are you able to supply me with an exact process to reproduce this? (Or an exact description -- is this a 2-of-3 using keys A, B, C, where both A and B have already signed, and you have C in your wallet, so it should be signable?)

Sjors commented:
It's a watch-only wallet with a 2-of-3 multisig (three hardware wallets). Two of them signed.

gwillen commented:
What is it misinterpreting? That is, what do you expect it to say? (If it's a watch-only wallet, then "this wallet does not have the right keys" should really say "this wallet cannot make signatures", because it's supposed to check privateKeysDisabled() first.)

Sjors commented:
The Broadcast Transaction button should not be disabled. I can finalize it with the RPC and `testmempoolaccept` is happy.

gwillen commented:
Oh!! Sorry, of course. You're right, I will look into why it's doing that. For testing broadcastability, I am relying on FillPSBT's "complete" output, which I thought was the right thing and should just work in this case. The actual text does rely on AnalyzePSBT, but the button should in theory still be enabled. I'll figure that out.

Sjors commented:
Mmm, actually it looks like I was testing on an outdated descriptor wallet branch (with your commits cherry-picked). I just tried on slightly more recent version and it now it recognized it as finished. So I think you're right that `FillPSBT` did the job (it's also a finaliser).

gwillen commented:
Oh, fantastic. Yeah, I have essentially ended up assuming that finalization is trivial and will always succeed given sufficient signatures, and I automatically attempt it before doing anything else.

jonasschnelli commented:
Concept ACK

instagibbs commented:
concept ACK, will start reviewing in depth post-merge of https://github.com/bitcoin/bitcoin/pull/17509

instagibbs commented:
from a test run:
```
* Sends 47.99999859 BTC to bcrt1q6yp5np47xjpcn06fxmut4rug5nll4uh4xmefwn
* Sends 2.00000000 BTC to bcrt1qr5nel4sx76q953crm7vq3qjd6umgyzmlufyfaa
```

Do we really want to display change outputs? I'm guessing not.

instagibbs commented:
now is a good time to rebase since parent PR is likely merged without change in April

gwillen commented:
Tests segfault on my OS X machine, but it seems that they do that even if I undo my changes, so not going to worry about it too much for the moment...

Needed some fixups to make it build after rebase, which I am at this moment too lazy to rebase into their proper places in the list of commits, but will do that later before this is ready for merge.

Sjors commented:
Travis macOS build is happy, as well as on my own machine (macOS 10.15.3)

instagibbs commented:
Try a `make clean` cycle. I've had really bizarre db-related errors that go away after that.

jb55 commented:
Concept ACK. With this branch I no longer need to hit the command line to finalize a PSBT! Although I still need HWI to sign :(

It might be useful to mention if the transaction is already in the mempool or a block? I noticed it always says broadcast even if it already was broadcasted. Happy to defer these improvements to future PRs, along with change detection ;)

jb55 commented:
Another thing I noticed is that when I have the wrong wallet selected I get conflicting messages:

![Apr14-142210](https://user-images.githubusercontent.com/45598/79275625-6469cb00-7e5b-11ea-94bf-8ce9a5043094.png)

"is fully signed and ready for broadcast"

and 

"Transaction as 1 unsigned inputs"

but this goes away with the right wallet selected:

![Apr14-142431](https://user-images.githubusercontent.com/45598/79275825-c0345400-7e5b-11ea-89de-89099916a551.png)

instagibbs commented:
> Although I still need HWI to sign :(

There's HWI-QT! ;)

meshcollider commented:
This can be rebased now #17509 is merged :)

instagibbs commented:
I'd wait for rebase until #16528 is merged since it appears close

instagibbs commented:
and now is the actual time to rebase

instagibbs commented:
I could take on this PR to champion it if you're too busy?

gwillen committed:
Make lint-spelling.py happy

gwillen committed:
[gui] PSBT Operations Dialog (sign & broadcast)

Add a "PSBT Operations" dialog, reached from the "Load PSBT..." menu
item, giving options to sign or broadcast the loaded PSBT as
appropriate, as well as copying the result to the clipboard or saving
it to a file.

gwillen commented:
Rebased!

I think I still need to address comments from @jb55. I should probably also consider improving handling of change addresses (but I'd like to defer that to another PR if nobody objects.)

instagibbs commented:
Change address handling is something we can improve later

instagibbs commented:
restarted an unrelated travis failure

luke-jr commented on src/qt/psbtoperationsdialog.cpp:
```diff
@@ -0,0 +1,266 @@
+
+    bool complete;
+    size_t n_could_sign;
+    TransactionError err = m_wallet_model->wallet().fillPSBT(SIGHASH_ALL, false /* sign */, true /* bip32derivs */,  m_transaction_data, complete, &n_could_sign);
```
Nit: Extra space before m_transaction_data

gwillen commented:
@jb55 I'd be up for adding 'in mempool / in block' detection in a future PR. Can you elaborate on what you mean by "right wallet" / "wrong wallet"? In theory once the tx is fully signed, it shouldn't matter what wallet you're using. So I'm confused about what's happening there. Can you give me a more precise sequence of steps to reproduce?

Sjors commented on src/qt/walletview.cpp:
```diff
@@ -205,23 +207,37 @@ void WalletView::gotoVerifyMessageTab(QString addr)
+        }
+        std::ifstream in(filename.toLocal8Bit().data(), std::ios::binary);
+        std::stringstream sstr;
+        sstr << in.rdbuf();
```
nit: this refactor in how the file is read might deserve its own commit.

gwillen replied:
I'm not sure how much sense that change makes in isolation -- it was forced by the clipboard change, in which it is included, because when I lifted the definition of 'data' to an outer scope, I couldn't figure out how to avoid changing the approach at the same time (since I was no longer able to use a non-default constructor of std::string.)

gwillen replied:
... And actually the refactor isn't necessary, I can do `data = std::string(std::istreambuf_iterator<char>{in}, {});` and move-assignment should cause the copy to be elided so it's free. So I'll just change it to that.

Sjors commented on src/qt/bitcoingui.cpp:
```diff
@@ -321,7 +321,7 @@ void BitcoinGUI::createActions()
     verifyMessageAction = new QAction(tr("&Verify message..."), this);
     verifyMessageAction->setStatusTip(tr("Verify messages to ensure they were signed with specified Bitcoin addresses"));
-    m_load_psbt_action = new QAction(tr("Load PSBT..."), this);
+    m_load_psbt_action = new QAction(tr("&Load PSBT from file..."), this);
```
nit: this keyboard shortcut deserves its own commit

Sjors commented on src/qt/psbtoperationsdialog.cpp:
```diff
@@ -0,0 +1,266 @@
+        QStringList alternativeUnits;
+        for (const BitcoinUnits::Unit u : BitcoinUnits::availableUnits())
+        {
+            if(u != m_client_model->getOptionsModel()->getDisplayUnit())
```
nit: brackets

gwillen replied:
good call, fixed

Sjors reviewed (approved):
tACK 7169451646511964f5356a8859f696b827dabe50 modulo rebase

I tested with a fairly simple setup (single signature Trezor wallet and HWI). I might take some followups to make it completely idiot proof and handle multi-sig with perfect messaging.

gwillen commented:
Rebasing across #19176, I turned a few instances of `QString::fromStdString(TransactionErrorString(err))` into `QString::fromStdString(TransactionErrorString(err).translated)` to get things to build -- I think that's the right approach but let me know if it's not.

gwillen commented:
It seems like @jb55's issue is consistent with FillPSBT not finalizing the transaction like I expected that it would, variably depending on which wallet is loaded. (@jb55, is one or both of the wallets a descriptor wallet by any chance?) `FinalizePSBT` calls `SignPSBTInput(DUMMY_SIGNING_PROVIDER, psbtx, i, SIGHASH_ALL);` on every input, whereas `FillPSBT` seems to bottom out in calls to `SignPSBTInput(HidingSigningProvider([some signing provider], !sign, !bip32derivs), psbtx, i, sighash_type);`. I would have hoped those would ultimately do the same thing, but this code is twisty, dark and deep... (And I'm nervous about the way that the ScriptPubKeyMan interface results in a somewhat complex web of signing logic getting semi-duplicated across each SPKM implementation. The wallet signing logic was just about beyond my comprehension already, before the addition of another layer.)

There are really two issues: (1) we're trying to finalize the transaction when we load it, so that we know whether it's ready or not, but failing. (2) our two information sources are not in sync (one of them we are calculating directly, the other we are getting from AnalyzePSBT, which appears to do its own finalization when checking if all signatures are present.)

I think I called FillPSBT because I wanted to know whether it thinks the PSBT is `complete`, and I assumed it would finalize in the process (and I think this might have been true before the SPKM refactor, though I'm not really very sure.)

So I think the fix to (1) is for me to add a call to FinalizePSBT in openWithPSBT before I call fillPSBT, which I have just done. And as for (2) I think it's okay to ignore it for now.

The fix is harmless even if I'm wrong, so I think it's okay to call it good at this point. But I would still really really like a repro test case to play with, @jb55. (If I have correctly deduced fillPSBT's behavior, I find it surprising.)

Can we officially declare this PR done and ready (and pray that this is the last time it will need a rebase?) :-)

instagibbs commented:
concept ACK fix, going to test again

instagibbs commented:
tested ACK https://github.com/bitcoin/bitcoin/pull/18027/commits/931dd4760855e036c176a23ec2de367c460e4243

Sjors commented:
re-tACK 931dd4760855e036c176a23ec2de367c460e4243

jonatack commented:
Concept ACK

achow101 reviewed (commented):
ACK 931dd4760855e036c176a23ec2de367c460e4243

gwillen commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -617,6 +620,14 @@ TransactionError LegacyScriptPubKeyMan::FillPSBT(PartiallySignedTransaction& psb
         SignPSBTInput(HidingSigningProvider(this, !sign, !bip32derivs), psbtx, i, sighash_type);
+
+        bool signed_one = PSBTInputSigned(input);
+        if (n_signed && (signed_one || !sign)) {
```
I'm very interested in a conversation about refactoring some of this stuff. Ok to save for another PR?

gwillen replied:
(I am interested both in (1) reducing code duplication and (2) working out more precisely whether we can sign something, which I found to be very difficult to actually compute, and I definitely recognize this as a conservative approximation.)

achow101 replied:
> Ok to save for another PR?

Yep

jb55 commented:
ACK 931dd4760855e036c176a23ec2de367c460e4243

meshcollider commented:
Concept ACK, glanced at the code, looks RTM

jonatack commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -617,6 +620,14 @@ TransactionError LegacyScriptPubKeyMan::FillPSBT(PartiallySignedTransaction& psb
+        bool signed_one = PSBTInputSigned(input);
+        if (n_signed && (signed_one || !sign)) {
+            // If sign is false, we assume that we _could_ sign if we get here. This
+            // will never have false negatives; it is hard to tell under what i
```
nit: s/what i/what/ here and at line 2136 below

jonatack commented on src/qt/walletview.cpp:
```diff
@@ -11,6 +11,7 @@
 #include <qt/askpassphrasedialog.h>
 #include <qt/clientmodel.h>
 #include <qt/guiutil.h>
+#include <qt/psbtoperationsdialog.h>
```
nit: sort

jonatack reviewed (commented):
Light code review and partially tested ACK 931dd4760855e036c176a23ec2de367c460e4243

Good to see this move forward and looking forward to continued progress as described in https://github.com/bitcoin/bitcoin/pull/18027#issuecomment-646533587. A couple of nits for the next PR follow, no need to respond to them.
